library/Vanilla/Authenticator/Authenticator.php

```diff
@@ -0,0 +1,74 @@
+<?php
+/**
+ * @author Alexandre (DaazKu) Chouinard <alexandre.c@vanillaforums.com>
+ * @copyright 2009-2017 Vanilla Forums Inc.
+ * @license http://www.opensource.org/licenses/gpl-2.0.php GNU GPL v2
+ */
+
+namespace Vanilla\Authenticator;
+
+use Garden\Web\RequestInterface;
+
+abstract class Authenticator {
+
+    /**
+     * Identifier of this authenticator instance.
+     *
+     * Extending classes will most likely require to have a dependency on RequestInterface so that they can
+     * fetch the ID from the URL and throw an exception if it is not found or invalid.
+     *
+     * @var string
+     */
+    private $authenticatorID;
+
+    /**
+     * Authenticator constructor.
+     *
+     * @throws \Exception
+     * @param string $authenticatorID
+     */
+    public function __construct($authenticatorID) {
+        $this->authenticatorID = $authenticatorID;
+
+        if (array_pop(explode('\\', static::class)) !== $this->getNameImpl(static::class).'Authenticator') {
+            throw new \Exception('Authenticator class name must end with Authenticator');
+        }
+    }
+
+    /**
+     * Validate an authentication by using the request's data.
+     *
+     * @throw Exception Reason why the authentication failed.
+     * @param RequestInterface $request
+     * @return array The user's information.
+     */
+    public abstract function validateAuthentication(RequestInterface $request);
+
+    /**
+     * Getter of the authenticator's ID.
+     *
+     * @return string
+     */
+    public final function getID() {
+        return $this->authenticatorID;
+    }
+
+    /**
+     * Default getName implementation.
+     *
+     * @return string
+     */
+    private function getNameImpl() {
```

Comment:
Is this necessary as its own function? Couldn't this just be the body of `getName` and child classes could overwrite it as necessary? If they really needed this exact routine, they could call the parent's method.

Reply:
`getNameImpl()` is also used in the constructor ;)